Respecting GitHub rate limits

shalevohad opened the pull request:
should fix #878

shalevohad committed:
Created new GitHubUpdater BaseClass

shalevohad committed:
Created new GitHubUpdater BaseClass

shalevohad committed:
make UpdateChecker inherit the GitHubUpdaterBase class and using base class rate-limit functions to project github

shalevohad committed:
make UpdateChecker inherit the GitHubUpdaterBase class and using base class rate-limit functions to project github

shalevohad committed:
Removing unnecessary threading - threading checking only takes place in the base

shalevohad committed:
Removing unnecessary threading - threading checking only takes place in the base

shalevohad committed:
make GithubUpdater class a static class (rename it from GitHubUpdaterBase to reflect the change in implementation)

shalevohad committed:
make GithubUpdater class a static class (rename it from GitHubUpdaterBase to reflect the change in implementation)

shalevohad committed:
use the new static class instead of the base class

shalevohad committed:
use the new static class instead of the base class

shalevohad committed:
cleaning code

shalevohad committed:
cleaning code

shalevohad committed:
Aligning the target framework with the other projects in the solution and adding a reference to the common project for the GitHubUpdater class

shalevohad committed:
Aligning the target framework with the other projects in the solution and adding a reference to the common project for the GitHubUpdater class

shalevohad committed:
Using GitHubUpdater to implement github rate limiter in auto update

shalevohad committed:
Using GitHubUpdater to implement github rate limiter in auto update

shalevohad committed:
Adding an option for a callback function when waiting due to maxRetries (mainly for the purpose of displaying status to the user from the calling function)

shalevohad committed:
Adding an option for a callback function when waiting due to maxRetries (mainly for the purpose of displaying status to the user from the calling function)

shalevohad committed:
add maxRetries function propertu to GetAllReleasesAsync function

shalevohad committed:
add maxRetries function propertu to GetAllReleasesAsync function

shalevohad committed:
Add rate limiter waiting status to AutoUpdater status label

shalevohad committed:
Add rate limiter waiting status to AutoUpdater status label

shalevohad committed:
Executing cancellationToken by clicking the cancel button to stop waiting for the git to retry

shalevohad committed:
Executing cancellationToken by clicking the cancel button to stop waiting for the git to retry

shalevohad committed:
Merge branch 'master' into Respecting-Github-rate-limits

gcask committed:
Only trace failed json deserialization on network messages.

Some servers can be subject to various bot probing, leading to a lot of log spamming.

Downgraded the error to a trace, as to give at least a way to diagnose them without needing an exe re-release.

gcask committed:
Merge pull request #908 from gcask/json-trace

Only trace failed json deserialization on network messages.

gcask committed:
Update F5E.lua

gcask committed:
Merge pull request #909 from gcask/fix-f5e

Update F5E.lua

ciribob committed:
Remove MIDS ambient audio for now

ciribob committed:
Merge branch 'release' into feature/instructor-mode

gcask committed:
Apply Volume radio on wet/dry overall mix

Fixes volume knob not working when using full dry.

gcask committed:
Ensure wet and dry don't provision from the same buffer (avoids double reads)

In line with the TX counterpart.
Also updated the system to use a Span as its output.

gcask committed:
Merge pull request #912 from gcask/fix-wetdry-volume

Fix wet/dry volume

gcask committed:
Wipe data on nil device.

gcask committed:
Merge pull request #913 from gcask/ejection-handling

Wipe data on nil device.

gcask committed:
Minor changes to logging

gcask committed:
Merge pull request #914 from gcask/debug-srs-gamegui

Minor logging tweaks

ciribob committed:
WIP

WIP of Intercom Monitoring

TODO - add a setting to enable or disable instructor mode
TODO test it!

ciribob committed:
release version 2.3.2.2

ciribob committed:
Lower ambient background cockpit noise

ciribob committed:
Revert to program files

ciribob committed:
Merge branch 'release' into feature/instructor-mode

suns8957 committed:
Update Modules\F14.lua for F14 Early support

gcask committed:
Merge pull request #916 from suns8957/master

Update Modules\F14.lua for F14 Early support

gcask committed:
Merge pull request #918 from ciribob/release

Release -> master

gcask committed:
Replace MessageBox in App with TaskDialog.

gcask committed:
Change singleton check to rely on Mutex

gcask committed:
Merge pull request #919 from gcask/917-singleton-mutex

917 singleton mutex

gcask committed:
Improve UDP handling

* Better thread-safety
* Simplified logic
* Less threads

ciribob committed:
Enable mute on the client window

Can now mute a client via the client window if you wish
First pass of instructor mode completed - need to add a client setting to enable or disable

ciribob committed:
Bump Version

Bump version for the new features

gcask committed:
Merge pull request #921 from ciribob/release

Release -> master

gcask committed:
Merge remote-tracking branch 'upstream/master' into improve-udp-handling

gcask committed:
Reset models on error

ciribob committed:
Merge branch 'release' into feature/instructor-mode

ciribob committed:
Update string

Cephy314 committed:
fix: renamed realRadioTx to realRadioRx

Name was wrong for RealRadioRx Option, was using the same name as Tx and causing duplicate match errors from the Options provider.

MrSuperjaffa committed:
Update C130J SAT Radio

gcask committed:
Update tasking logic

Make it clearer, more resilient

gcask committed:
Extra logging on the server when it doesn't find a client.

gcask committed:
Merge pull request #924 from MrSuperjaffa/c130j-update

Update C130J SAT Radio

gcask committed:
Ensure sources access is always under lock

gcask committed:
Update RadioMixingProvider.cs

gcask committed:
Correctly reset the ping task on timeout.

gcask committed:
Switch to task-based

gcask committed:
Change UDPVoiceRouter to be task-based

gcask committed:
Merge remote-tracking branch 'upstream/master' into improve-udp-handling

gcask committed:
Batch drain

gcask committed:
Merge pull request #923 from Cephy314/fix/cmdln-server-properties

fix: renamed realRadioTx to realRadioRx

gcask committed:
Cancellation token pass.

Mainly ensuring they are properly reset to a new state, instead of using cancelled ones.

gcask committed:
Add support for F16 U/VHF DED activity indicator

gcask committed:
Merge pull request #927 from gcask/f16-activity-indicator

Add support for F16 U/VHF DED activity indicator

gcask committed:
Merge pull request #920 from gcask/improve-udp-handling

Improve UDP handling

gcask committed:
Merge pull request #928 from ciribob/master

Master -> Beta

gcask committed:
Merge branch 'beta' into feature/instructor-mode

gcask committed:
Update InstructorModeViewModel.cs

Fix typo

ciribob committed:
Merge pull request #926 from ciribob/feature/instructor-mode

add new instructor mode

gcask committed:
Ensure filters have unique instance per radio models

gcask committed:
Merge pull request #930 from gcask/filters-cloning-fixes

Ensure filters have unique instance per radio models

gcask committed:
Merge pull request #931 from ciribob/master

Update Beta

gcask committed:
Handle lack of radio info

ciribob committed:
DIS Gateway changes

TBC - needs review

ciribob committed:
add unit ID

ciribob committed:
Bump version

ciribob committed:
Merge branch 'beta' into feature/dis-gateway

ciribob committed:
Handle disconnected gateway clients

ciribob committed:
version bump

ciribob committed:
Merge branch 'beta' into feature/dis-gateway

ciribob committed:
Pin SDK version

gcask committed:
Updating radio device IDs for the CH47F.

gcask committed:
Merge pull request #932 from gcask/ch-47f-2.9.22.17790-fix

Updating radio device IDs for the CH47F.

gcask committed:
Merge pull request #933 from ciribob/master

master -> beta

gcask committed:
Add more logging around disconnect and server switching.

gcask committed:
Reintroduce synchronized method

gcask committed:
Introduced PooledArray struct for rented arrays.

gcask committed:
Fix copying audio to the wrong offset

gcask committed:
Better resiliency

* Opted out of pooled arrays for non local items.
* More aggressive clamping and checking.
* Defensive code against empty arrays.

gcask committed:
Fix null check

gcask committed:
Handle decode errors at lower level

gcask committed:
Apply CancellationToken treatment to TCPClientHandler

gcask committed:
Merge pull request #934 from gcask/extra-logging

Extra logging, improve resiliency

gcask committed:
Merge pull request #935 from ciribob/beta

Beta to Release

gcask committed:
Slight tweaks in how verbose exceptions are.

gcask committed:
Merge pull request #936 from gcask/fix-verbose-cancel

Slight tweaks in how verbose exceptions are.

ciribob committed:
Fix for external audio

UnitID is actually optional now
Record is respected - defaults to true

ciribob committed:
Merge branch 'release' into beta

ciribob committed:
Version bump 2.3.3.3

ciribob committed:
Fix build with .net 10 installed

ciribob committed:
Bump version

gcask committed:
Upgrade projects to .NET10

gcask committed:
dependency updates

gcask committed:
Update workflow to .NET 10

gcask committed:
Upgrades to github workflow.

gcask committed:
Attempt to use server image.

gcask committed:
Update dependencies.

gcask committed:
Update workflow

gcask committed:
Merge pull request #938 from gcask/update-dependencies

Update dependencies

gcask committed:
merge upstream/master

gcask committed:
Merge remote-tracking branch 'upstream/master' into speex-jitter

ciribob committed:
Finish DIS support

ciribob committed:
Merge branch 'beta' into feature/dis-gateway

ciribob committed:
Merge branch 'release' into feature/dis-gateway

ciribob committed:
Force background ambient audio to never blow the pipeline of 1,-1

Force ambient audio to always be 0.25 if for whatever reason its not parsed correctly

ciribob committed:
Update ClientAudioProvider.cs

ciribob committed:
Use equalAny

ciribob committed:
Remove mock

ciribob committed:
Change the ambient audio to instead be a target we're trying to hit

Calculate RMS for an effect

Then aim for a % loudness of that effect, first calculate the dB diff - then back to linear

ciribob committed:
Change to keep it vectorised

ciribob committed:
Possible fix for WINE users

ciribob committed:
Remove Mic AGC

ciribob committed:
Merge pull request #946 from ciribob/feature/dis-gateway

Feature/dis gateway

ciribob committed:
Remove commented out bits - thats what VC is for

ciribob committed:
Merge pull request #945 from ciribob/remove-mic-agc

Remove Microphone AGC

ciribob committed:
Merge branch 'beta' into ambient-audio-bug-fix

ciribob committed:
Fix for linux

Force software resampler by pretending we're WindowsN

ciribob committed:
Merge pull request #944 from ciribob/ambient-audio-bug-fix

Possible Audio Loss Fix

ciribob committed:
Version bump

ciribob committed:
Merge branch 'master' into beta

ciribob committed:
Fix library incompatibility

Microsoft.Windows.Compatibility version must match .net version (currently 9)

Clean up leftover debug code

gcask committed:
Initial support for AH-6J/MH-6J

gcask committed:
Merge pull request #948 from gcask/ah6j

Initial support for AH-6J/MH-6J

gcask committed:
Update runner to 2026.

gcask committed:
Merge remote-tracking branch 'upstream/master' into dotnet10

gcask committed:
Merge pull request #953 from ciribob/beta

Beta -> release

gcask committed:
Merge pull request #954 from ciribob/release

Release -> master

gcask committed:
Add configurable address for http REST server.
Default it to 'localhost'

gcask committed:
Use setting-driven address

gcask committed:
Add http server address as option.

gcask committed:
Merge remote-tracking branch 'upstream/master' into dotnet10

gcask committed:
Update .NET version in global.json

gcask committed:
Light should only be active on TX, not RX in  the F16.

fixes #943

gcask committed:
specify version for tests

gcask committed:
Merge pull request #955 from gcask/#943-f16-activity-light

Light should only be active on TX, not RX in  the F16.

gcask committed:
Moved some logs (namely mods lookup) to debug level.

gcask committed:
Merge pull request #956 from gcask/reduced-log-verbosity

Moved some logs (namely mods lookup) to debug level.

gcask committed:
Merge pull request #957 from ciribob/master

master -> release

gcask committed:
Bump min windows version to 10.0.22621.0 (22H2)
Bump to VS 2026

gcask committed:
Update self contained flags for publish script.

gcask committed:
Merge pull request #937 from gcask/dotnet10

.NET 10

gcask committed:
Merge remote-tracking branch 'upstream/master' into speex-jitter

gcask committed:
Merge remote-tracking branch 'upstream/master' into speex-jitter

gcask committed:
Add port of Speex's JitterBuffer.

gcask committed:
Switch the provider interface to using the speex version.

gcask committed:
Removed native binding

gcask committed:
Renamed Jitter to JitterBuffer

gcask committed:
Fix Dispose pattern for speex processor

gcask committed:
Bump to 2.3.5.0

gcask committed:
Merge pull request #962 from gcask/release-2.3.5.0

Bump to 2.3.5.0

gcask committed:
Merge remote-tracking branch 'upstream/release' into release-to-master

gcask committed:
Merge pull request #961 from ciribob/master

master -> release

gcask committed:
2.3.6.0

gcask committed:
Merge branch 'release' of https://github.com/ciribob/DCS-SimpleRadioStandalone into delivery-master-to-release

gcask committed:
Merge pull request #960 from gcask/speex-jitter

Speex jitter

gcask committed:
2.3.7.0

gcask committed:
JitterBuffer port fixes

gcask committed:
Minor tweaks to the tick logic.

gcask committed:
Merge pull request #964 from gcask/jitter-buffer-fix

Jitter buffer fix

gcask committed:
2.3.7.1

gcask committed:
Removed forced C# version target and non nullable attribute.

gcask committed:
Removed unnecessary package references

gcask committed:
Updated compile target

gcask committed:
unreachable code in DEBUG warning.

gcask committed:
nullable reference context warning.

gcask committed:
bad define

gcask committed:
Assert.HasCount

gcask committed:
Deprecated function usage

gcask committed:
Merge pull request #965 from gcask/warnings-pass

Warnings pass

ciribob committed:
Update signtool path

ciribob committed:
Update signtool path

PoorCameraQuality committed:
fix: use DPI-independent bounds for window visibility

Made-with: Cursor

Commit:
feat: add transponder power toggle binding

Made-with: Cursor

gcask committed:
Merge pull request #970 from PoorCameraQuality/pr-628-transponder-toggle

feat: add transponder power toggle binding Feat #628

gcask commented on Common/Helpers/GitHubUpdater.cs:
```diff
@@ -0,0 +1,89 @@
+    private static readonly int DefaultMaxRetries = 3;
+
+    // Semaphore to serialize all update checks
+    private static readonly SemaphoreSlim UpdateSemaphore = new(1, 1);
```
I think that's overkill.
Let Github decide and tell each check their own limit.
Most of the time a single person will be able to run concurrent-ish update checks.

gcask commented on Common/Helpers/GitHubUpdater.cs:
```diff
@@ -0,0 +1,89 @@
+                    // Notify the caller about the wait
+                    onRateLimitWait?.Invoke(waitFor, attempt, maxRetries);
+
+                    await Task.Delay(waitFor, cancellationToken);
```
I would suggest kicking off the delay, then calling the user callback, then awaiting the delay.
This way however long the user callback happens to be, that doesn't count on top of the requested wait.

gcask commented on Common/Helpers/UpdaterChecker.cs:
```diff
@@ -10,54 +10,35 @@
-            {
-                if (_instance == null) _instance = new UpdaterChecker();
-            }
+    public static UpdaterChecker Instance => _instance;
```
If you're doing this - simple get only property set to the new instance will do nicely

gcask commented on Common/Helpers/UpdaterChecker.cs:
```diff
@@ -10,54 +10,35 @@
-            return _instance;
-        }
-    }
+    public UpdaterChecker() { }
```
should be private, otherwise anyone can create one and it's not much of a singleton :)

gcask commented on Common/Helpers/UpdaterChecker.cs:
```diff
@@ -10,54 +10,35 @@
-            var githubClient = new GitHubClient(new ProductHeaderValue(GITHUB_USER_AGENT, VERSION));
+//#if DEBUG
+//            return;
+//#endif
```
Restore, we don't really want update checks in DEBUG builds.

gcask commented on AutoUpdater/MainWindow.xaml.cs:
```diff
@@ -155,49 +154,80 @@ private Release FindRightRelease(IReadOnlyList<Release> releases)
+        catch (OperationCanceledException)
+        {
+            Status.Content = "Update cancelled due to rate limit.";
+            return null;
```
We might want to re-throw here instead, so the cancellation is correctly propagated upwards?

Commit:
fix: use VirtualLeft/VirtualTop for reset positions (gcask review)

Made-with: Cursor

gcask committed:
AssignButton non blocking.

gcask committed:
UDPCommandHandler main task non blocking.

gcask committed:
VAICOM main task non blocking.

gcask committed:
LotATCSyncHandler non blocking.

gcask committed:
autoconnect handler non blocking.

gcask committed:
gamegui handler non blocking.

gcask committed:
LOSHandler non blocking

gcask committed:
Radio sync handler non blocking.

gcask committed:
Changed some task from old factory new to simple run.

gcask committed:
Use cancellation token for UDP audio processor decode.

gcask committed:
Rollback some changes in radio sync handler.

gcask committed:
Handle TCP client connect gracefully through linked source tokens.

gcask committed:
Fix async/await for updater check.

gcask committed:
Update MainWindow.xaml.cs

gcask committed:
Make sure async Task is used instead of async void.
Using "Taskable" timer for TCPClientHandler.
Simplifying logic for UDPVoiceHandler

gcask committed:
merging upstream/beta

gcask committed:
sideload: fix exception when the -lang option was not provided.

gcask committed:
sideload: jitterbuffer handles packet smaller than expected gracefully.

gcask committed:
Skip client audio if we have a null provider.

gcask committed:
Merge pull request #971 from PoorCameraQuality/pr-312-dpi-window-visibility

Fix incorrect “window not visible” warning on high DPI setups #312

gcask committed:
Remove warning for deliberate async update check.

gcask committed:
Move min/max thread back behind define with a comment.

gcask committed:
Upgrade logging helper code.

gcask committed:
merge master

gcask committed:
Merge remote-tracking branch 'upstream/beta' into task-model

gcask committed:
Moved Server sync to be task-based.
Update logic in NATHandler to be task compatible.

gcask committed:
Avoid situation where we would 'await null'

gcask committed:
Tracking code for threading issues.

gcask committed:
Allow NATHandler setup to cancel after ten seconds overall, including creating the port mappings.

gcask committed:
Async pass on CLI Server

gcask committed:
ExternalAudioClient async pass

gcask committed:
Remove volatile requirement for _started

gcask committed:
Merge pull request #976 from gcask/task-model

Improve async/await usage + misc pass.

gcask committed:
2.3.8.0

Rob2816 committed:
UDP Commands Expansion

Rob2816 committed:
UDP Commands Expansion

Rob2816 committed:
UDP Commands Expansion

Rob2816 committed:
UDP Commands Expansion

gcask committed:
async logic for autoconnect handler message... handling.

gcask committed:
Update 'mismatch server' prompt to async TaskDialog.

Rob2816 committed:
Simul to Simultaneous; Chan to Channel.

Rob2816 committed:
Restored null check for UI update logic, accidentally moved during previous commit.

gcask committed:
More async usage.

itsmaffies committed:
Updated NatHandler.cs to support PMP requests

Small or statement to call PortMapper.Pmp

gcask committed:
Merge pull request #983 from gcask/autoconnect-async

Autoconnect async

gcask committed:
2.3.8.1

ciribob committed:
Switch to input saving on select

ciribob committed:
Async the dialog

ciribob committed:
No await void - bad

ciribob committed:
Add await

ciribob committed:
fancy

Task.Run(async () => await ShowWarningAsync()).Wait();

gcask committed:
Merge pull request #988 from ciribob/bugfix/save-input-on-change

Bugfix/save input on change

gcask committed:
2.3.8.2

gcask committed:
Merge remote-tracking branch 'upstream/beta' into delivery-master-to-release

gcask committed:
Merge branch 'master' into master

gcask committed:
Merge pull request #985 from itsmaffies/master

Updated NatHandler.cs to support PMP requests

gcask committed:
Bump dependencies.

gcask committed:
Merge pull request #989 from gcask/dependencies-update

Bump dependencies.

Rob2816 committed:
Code simplified using SetGlobalSimultaneousTransmission in ToggleGlobalSimultaneousTransmission

Rob2816 committed:
Code simplified using dcsPlayerRadioInfo instead of _clientStateSingleton.DcsPlayerRadioInfo

Rob2816 committed:
Merge branch 'ciribob:master' into udp-commands-expansion

Rob2816 committed:
Removed: using Ciribob.DCS.SimpleRadio.Standalone.Client.UI.ClientWindow.AwacsRadioOverlayWindow;

Rob2816 committed:
Merge branch 'udp-commands-expansion' of https://github.com/Rob2816/DCS-SimpleRadioStandalone-CO into udp-commands-expansion

gcask committed:
Merge pull request #978 from Rob2816/udp-commands-expansion

UDP Commands Expansion

gcask committed:
Update singleton.
Removed semaphore.

gcask committed:
Changed from callback to task-based approach.

gcask committed:
Update from MessageBox to TaskDialog.

gcask committed:
CLI server no longer waits on the update check.

gcask committed:
Merge branch 'Respecting-Github-rate-limits' of https://github.com/shalevohad/DCS-SimpleRadioStandalone into pr/884
